exp/api/v1alpha3/types.go

```diff
@@ -83,3 +83,131 @@ var (
+	// searching for AMIs that match the pattern capa-ami-ubuntu-?1.18.0-* for a
+	// Machine that is targeting kubernetes v1.18.0 and the ubuntu base OS. See
+	// also: https://golang.org/pkg/text/template/
+	// +optional
+	ImageLookupFormat string `json:"imageLookupFormat,omitempty"`
+
+	// ImageLookupOrg is the AWS Organization ID to use for image lookup if AMI is not set.
+	ImageLookupOrg string `json:"imageLookupOrg,omitempty"`
+
+	// ImageLookupBaseOS is the name of the base operating system to use for
+	// image lookup the AMI is not set.
+	ImageLookupBaseOS string `json:"imageLookupBaseOS,omitempty"`
+
+	// InstanceType is the type of instance to create. Example: m4.xlarge
+	InstanceType string `json:"instanceType,omitempty"`
+
+	// RootVolume encapsulates the configuration options for the root volume
+	// +optional
+	RootVolume *infrav1.Volume `json:"rootVolume,omitempty"`
+
+	// SSHKeyName is the name of the ssh key to attach to the instance. Valid values are empty string (do not use SSH keys), a valid SSH key name, or omitted (use the default SSH key name)
+	// +optional
+	SSHKeyName *string `json:"sshKeyName,omitempty"`
+
+	VersionNumber *int64 `json:"versionNumber,omitempty"`
+
+	// AdditionalSecurityGroups is an array of references to security groups that should be applied to the
+	// instances. These security groups would be set in addition to any security groups defined
+	// at the cluster level or in the actuator.
+	// +optional
+	AdditionalSecurityGroups []infrav1.AWSResourceReference `json:"additionalSecurityGroups,omitempty"`
+}
+
+// Overrides from describe-auto-scaling-groups
+type Overrides struct {
+	InstanceType string `json:"instanceType"`
+}
+
+// InstancesDistribution from describe-auto-scaling-groups
+type InstancesDistribution struct {
+	// +kubebuilder:validation:Enum=prioritized
+	// +kubebuilder:default=prioritized
+	OnDemandAllocationStrategy *string `json:"onDemandAllocationStrategy,omitempty"`
+
+	// +kubebuilder:validation:Enum=lowest-price;capacity-optimized
+	// +kubebuilder:default=lowest-price
+	SpotAllocationStrategy *string `json:"spotAllocationStrategy,omitempty"`
+
+	// +kubebuilder:default=0
+	OnDemandBaseCapacity *int64 `json:"onDemandBaseCapacity,omitempty"`
+
+	// +kubebuilder:default=100
+	OnDemandPercentageAboveBaseCapacity *int64 `json:"onDemandPercentageAboveBaseCapacity,omitempty"`
+}
+
+// MixedInstancesPolicy from describe-auto-scaling-groups
+type MixedInstancesPolicy struct {
+	InstancesDistribution *InstancesDistribution `json:"instancesDistribution,omitempty"`
+	Overrides             []Overrides            `json:"overrides,omitempty"`
+}
+
+// Tags
+type Tags map[string]string
+
+// AutoScalingGroup describes an AWS autoscaling group.
+type AutoScalingGroup struct {
+	// The tags associated with the instance.
+	ID              string            `json:"id,omitempty"`
+	Tags            map[string]string `json:"tags,omitempty"`
```

Comment:
Could we use `infrav1.Tags` instead of the `map[string]string`?

You may also need to make a change in the [tags package](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/tags/tags.go)....but hopefully the existing `WithEC2` will be ok

Reply:
it works like magic

Reply:
oh no. no it doesn't work. let me try fiddling with the tags package?